proposed/http-handlers/request-handlers-meta.md

```diff
@@ -368,6 +364,233 @@ executing that handler. If unable to route to a handler, it would instead
+middleware are provided to demonstrate how to write middleware that is
+loosely-coupled.
+
+Please note that these are not suggested as definitive or exclusive approaches
+to defining middleware dispatch systems.
+
+#### Queue-based request handler
+
+In this approach, a request handler maintains a queue of middleware, and a
+fallback response to return if the queue is exhausted without returning a
+response. When executing the first middleware, the queue passes itself as a
+request handler to the middleware.
+
+```php
+class QueueRequestHandler implements RequestHandlerInterface
+{
+    private $middleware = [];
+    private $fallbackHandler;
+    
+    public function __construct(RequestHandlerInterface $fallbackHandler)
+    {
+        $this->fallbackHandler = $fallbackHandler;
+    }
+    
+    public function add(MiddlewareInterface $middleware)
+    {
+        $this->middleware[] = $middleware;
+    }
+    
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        // Last middleware in the queue has called on the request handler.
+        if (0 === count($this->middleware)) {
+            return $this->fallbackHandler->handle($request);
+        }
+        
+        $middleware = array_shift($this->middleware);
+        return $middleware->process($request, $this);
+    }
+}
+```
+
+An application bootstrap might then look like this:
+
+```php
+// Fallback handler:
+$fallbackHandler = new NotFoundHandler();
+
+// Create request handler instance:
+$app = new QueueResponseHandler($fallbackHandler);
+
+// Add one or more middleware:
+$app->add(new AuthorizationMiddleware());
+$app->add(new RoutingMiddleware());
+
+// execute it:
+$response = $app->handle(ServerRequestFactory::fromGlobals());
+```
+
+This system has two request handlers: one that will produce a response if the
+last middleware delegates to the request handler, and one for dispatching the
+middleware layers. (In this example, the `RoutingMiddleware` will likely execute
+composed handlers on a successful route match; see more on that below.)
+
+This approach has the following benefits:
+
+- Middleware does not need to know anything about any other middleware or how it
+  is composed in the application.
+- The `QueueRequestHandler` is agnostic of the PSR-7 implementation in use.
+- Middleware is executed in the order it is added to the application, making the
+  code explicit.
+- Generation of the "fallback" response is delegated to the application
+  developer. This allows the developer to determine whether that should be a
+  "404 Not Found" condition, a default page, etc.
+
+#### Decoration-based request handler
+
+In this approach, a request handler implementation decorates both a middleware
+instance and a fallback request handler to pass to it. The application is built
+from the outside-in, passing each request handler "layer" to the next outer one.
+
+```php
+class DecoratedRequestHandler implements RequestHandlerInterface
```

Comment:
Nitpick: Technically, this isn't a decoratED handler.  It's a decoratING handler.  The name should be more descriptive.  Sorry for not catching that before.

Reply:
Fixed.